### DISCUSSION TITLE
Compiled Bindings : $parent.DataContext not typed correctly and does not bind to methods

### DISCUSSION THREAD

**Discussion by LelouBil** [AvaloniaUI/Avalonia#11962] [Q&A] [answered]
**Describe the bug**
$parent.DataContext does not correctly types the DataContext

**To Reproduce**

1. Use the sample MVVM App
2. Add the following to MainWindowViewModel.cs
```c#
 public ObservableCollection<string> Data { get; } = new() { "1", "2", "3" };
    public ReactiveCommand<string,Unit> LogDataCommand { get; } = ReactiveCommand.Create<string,Unit>(data =>
    {
       Debug.WriteLine("Data: " + data);
        return Unit.Default;
    });
```
4. Add the following to MainWindow.axaml
```xaml
<ItemsControl ItemsSource="{Binding Data}">
            <ItemsControl.DataTemplates>
                <DataTemplate x:DataType="system:String">
                    <Button CommandParameter="{Binding .}" Command="{Binding $parent[ItemsControl].DataContext.LogDataCommand}" Margin="0,5"><TextBlock Text="{Binding .}"></TextBlock></Button>
                </DataTemplate>
            </ItemsControl.DataTemplates>
        </ItemsControl>
```
5. A compilation error occurs
`[AVLN:0004] Unable to resolve property or method of name 'LogDataCommand' on type 'System.Object'.`

**Expected behavior**
I expect the code to compile and the buttons to be enabled. I also expect to see "Data : <number" in the debug log when I click on one of the buttons.

**Desktop :**

- OS: Windows 10
- Version 11.0.0-rc1.1

**Additional context**
Adding `x:CompileBindings="False"` makes the code compile and *work correctly* and turns the error into a warning.
However when trying to bind to a method, even disabling compiled bindings does not make it work

**Comment by timunie** [comment, marked as answer]
Please read the docs for compiled bindings carefully. https://docs.avaloniaui.net/docs/next/basics/data/data-binding/compiled-bindings#type-casting

**Reply by DevAM-Tools** [reply]
Thr casting hint should also be included here: https://docs.avaloniaui.net/docs/guides/data-binding/binding-to-controls#binding-to-an-ancestor

**Reply by timunie** [reply]
that is only needed in case of CompiledBindings. But as this is enabled by default now, you are right. Could you file a PR that adds a link to the cast section?

**Reply by DevAM-Tools** [reply]
I am willing to do that but that would be my first PR and I don't know how to do? Is there an explanation or tutorial available?

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/avalonia-docs?tab=readme-ov-file#contributing

**Reply by DevAM-Tools** [reply]
My first attempt: https://github.com/AvaloniaUI/avalonia-docs/pull/381

**Comment by timunie** [comment]
As I'm pretty sure with casting it will work, I'll move this into our Q&A section.
